// @agconnect/cloudstorage-server/dist/utils/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateCorsConfig = exports.validateBucketMetadata = exports.validateFileName = exports.validateDestinationString = exports.validateBucketNameString = exports.validateBucketName = void 0;
var error_1 = require("../error/error");
function validateBucketName(func, name) {
    var regex = /^[a-z0-9][a-z0-9-]{1,55}[a-z0-9]$/;
    if (!regex.test(name)) {
        throw error_1.invalidBucket(func);
    }
}
exports.validateBucketName = validateBucketName;
function validateBucketNameString(func, name) {
    var regex = /^[a-z0-9][a-z0-9-]{1,61}[a-z0-9]$/;
    if (!regex.test(name)) {
        throw error_1.invalidBucketString(func);
    }
}
exports.validateBucketNameString = validateBucketNameString;
function validateDestinationString(func, des) {
    if (!validateStorageObject(des)) {
        throw error_1.invalidDestinationString(func);
    }
}
exports.validateDestinationString = validateDestinationString;
function validateFileName(func, name) {
    if (!validateStorageObject(name)) {
        throw error_1.invalidFileString(func);
    }
}
exports.validateFileName = validateFileName;
function validateBucketMetadata(func, metadata) {
    if (!metadata || !metadata.cors) {
        throw error_1.invalidBucketMetadata(func);
    }
}
exports.validateBucketMetadata = validateBucketMetadata;
function validateCorsConfig(func, config) {
    if (!config) {
        throw error_1.invalidCorsConfig(func);
    }
}
exports.validateCorsConfig = validateCorsConfig;
function validateStorageObject(name) {
    var str = '#*:?\'"<>|[]';
    for (var _i = 0, name_1 = name; _i < name_1.length; _i++) {
        var item = name_1[_i];
        if (str.indexOf(item) !== -1) {
            return false;
        }
    }
    return true;
}
