// @agconnect/database-server/dist/exception/CloudDBErrorCode.js
"use strict";
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2020. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudDBErrorCode = void 0;
/**
 * Cloud DB error Code
 *
 * @author linhongyi [linhongyi2@huawei.com]
 * @since 2020-10-30
 */
var CloudDBErrorCode;
(function (CloudDBErrorCode) {
    CloudDBErrorCode["INPUT_IS_NULL"] = "1001";
    CloudDBErrorCode["DATA_SIZE_OVERFLOW"] = "1003";
    CloudDBErrorCode["QUERY_CONDITION_IS_NULL"] = "1006";
    CloudDBErrorCode["QUERY_INPUT_FIELD_NAME_IS_NULL"] = "1011";
    CloudDBErrorCode["QUERY_INPUT_FIELD_NAME_DO_NOT_EXIST"] = "1012";
    CloudDBErrorCode["INPUT_LIMIT_OR_OFFSET_LESS_THAN_ZERO"] = "1014";
    CloudDBErrorCode["QUERY_INPUT_VALUE_IS_NULL"] = "1015";
    CloudDBErrorCode["AGGREGATE_QUERY_CLOUDDB_ZONE_IS_NULL"] = "1018";
    CloudDBErrorCode["AGGREGATE_QUERY_FIELD_NAME_IS_NULL"] = "1019";
    CloudDBErrorCode["AGGREGATE_QUERY_FIELD_TYPE_IS_NOT_NUMERIC"] = "1020";
    CloudDBErrorCode["AGGREGATE_QUERY_FIELD_NAME_DO_NOT_EXIST"] = "1021";
    CloudDBErrorCode["QUERY_INPUT_QUERY_OBJECT_IS_NULL"] = "1022";
    CloudDBErrorCode["PAGINATE_QUERY_INPUT_IS_MORE_THAN_ONE"] = "1023";
    CloudDBErrorCode["PAGINATE_QUERY_INPUT_IS_ONLY_SUPPORTED"] = "1024";
    CloudDBErrorCode["PAGINATE_EQUAL_TO_NOT_SUPPORT_PRIMARY_KEY"] = "1025";
    CloudDBErrorCode["PAGINATE_ORDER_BY_NOT_SUPPORT_TYPE"] = "1026";
    CloudDBErrorCode["PAGINATE_EQUAL_TO_FIELD_IS_SAME_WITH_ORDER_BY"] = "1027";
    CloudDBErrorCode["FAIL_TO_PARSE_QUERY_DATA"] = "1035";
    CloudDBErrorCode["PAGINATE_NOT_SUPPORT_AGGREGATE"] = "1036";
    CloudDBErrorCode["EQUAL_TO_NOT_BEFORE_ORDER_BY"] = "1038";
    CloudDBErrorCode["PAGINATE_EQUAL_TO_NOT_SUPPORT_TEXT"] = "1039";
    CloudDBErrorCode["ORDER_BY_ONLY_SUPPORT_ONE_DIRECTION"] = "1040";
    CloudDBErrorCode["PAGINATION_ORDER_BY_HAS_DUPLICATE_FIELD"] = "1042";
    CloudDBErrorCode["DO_NOT_SUPPORT_BYTEARRAY"] = "1043";
    CloudDBErrorCode["DO_NOT_SUPPORT_BYTEARRAY_OR_BOOLEAN"] = "1044";
    CloudDBErrorCode["ONLY_SUPPORT_TEXT_AND_STRING"] = "1045";
    CloudDBErrorCode["NUMBER_IS_INVALID"] = "1046";
    CloudDBErrorCode["CLASS_INVALID"] = "1047";
    CloudDBErrorCode["FAIL_TO_GET_CREDENTIAL"] = "1048";
    CloudDBErrorCode["RESPONSE_IS_INVALID"] = "1049";
    CloudDBErrorCode["ENCRYPTED_FIELD_NOT_SUPPORT_QUERY"] = "1050";
    CloudDBErrorCode["USERID_IS_INVALID"] = "1051";
    CloudDBErrorCode["OBJECT_IS_INVALID"] = "1052";
    CloudDBErrorCode["ORDERBY_DO_NOT_SUPPORT_TEXT_OR_BYTEARRAY"] = "1053";
    CloudDBErrorCode["DATE_STRING_INVALID"] = "1055";
    CloudDBErrorCode["AGCONNECT_CLOUDDB_NOT_INITIALIZE"] = "1056";
    CloudDBErrorCode["DATABASE_NAME_INVALID"] = "1057";
    CloudDBErrorCode["ONLY_SUPPORT_ONE_SCHEMA"] = "1058";
    CloudDBErrorCode["INPUT_CONTAIN_ILLEGAL_CHARACTER"] = "1059";
    /**
     * Parameter invalid
     * User check the parameter corresponding to his operation.
     */
    CloudDBErrorCode["PARAMETER_INVALID"] = "1070008";
    /**
     * Data invalid, include data type mismatch and data overflow
     * User modify the data.
     */
    CloudDBErrorCode["ERROR_CODE_DATA_INVALID"] = "1070007";
    /**
     * Schema not exist
     * User need to check the schema verification
     */
    CloudDBErrorCode["SCHEMA_NOT_EXISTING"] = "1070012";
    /**
     * Data does not conclude primary key
     * User need to modify the data
     */
    CloudDBErrorCode["ERROR_CODE_DATA_NOT_CONCLUDE_PRIMARY_KEY"] = "1370003";
    /** A maximum of 1000 data records within 20 MB can be queried at a time
     * Adjust the query condition.
     */
    CloudDBErrorCode["ERROR_QUERY_DATA_TOO_LARGE"] = "1370019";
    /** The transaction failed because of the version of object verified failed.
     * User need to retry.
     */
    CloudDBErrorCode["ERROR_TRANSACTION_FAILED_DATA_CHANGED"] = "1370022";
    /**
     * The query condition does not match the index.
     * User need to modify the query condition.
     */
    CloudDBErrorCode["ERROR_CODE_PAGINATION_QUERY_FAILED_FOR_NO_INDEX"] = "1370029";
    /**
     * The transaction operation sequence is incorrect.
     * User need to modify the query condition.
     */
    CloudDBErrorCode["ERROR_CODE_TRANSACTION_SEQUENCE_INCORRECT"] = "1370035";
})(CloudDBErrorCode = exports.CloudDBErrorCode || (exports.CloudDBErrorCode = {}));
