// @agconnect/database-server/dist/utils/DataModelHelper.js
"use strict";
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2020. All rights reserved.
 */
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataModelHelper = exports.Validators = exports.Calculators = void 0;
var Utils_1 = require("./Utils");
var SchemaUtils_1 = require("./SchemaUtils");
var EMPTY_OBJECT_MEMORY_USAGE = 16;
var INTEGER_MIN_VALUE = -2147483648;
var INTEGER_MAX_VALUE = 2147483647;
var SHORT_MIN_VALUE = -32768;
var SHORT_MAX_VALUE = 32767;
var BYTE_MIN_VALUE = -128;
var BYTE_MAX_VALUE = 127;
var FLOAT_MIN_VALUE = (-(3.402823e+38));
var FLOAT_MAX_VALUE = 3.402823e+38;
var MAX_STRING_LENGTH = 200;
exports.Calculators = {};
function addCalculator(type, calculator) {
    exports.Calculators[type] = calculator;
}
function calculateStringUtf8Length(str) {
    var length = 0;
    if (!str) {
        return length;
    }
    for (var i = 0; i < str.length; i++) {
        var code = str.charCodeAt(i);
        length += code >> 11 ? 3 : code >> 7 ? 2 : 1;
    }
    return length;
}
addCalculator(SchemaUtils_1.FieldType.Integer, function (_) { return 4; });
addCalculator(SchemaUtils_1.FieldType.String, function (value) {
    return calculateStringUtf8Length(value);
});
addCalculator(SchemaUtils_1.FieldType.Short, function (_) { return 2; });
addCalculator(SchemaUtils_1.FieldType.Long, function (_) { return 8; });
addCalculator(SchemaUtils_1.FieldType.Boolean, function (_) { return 1; });
addCalculator(SchemaUtils_1.FieldType.Byte, function (_) { return 1; });
addCalculator(SchemaUtils_1.FieldType.Float, function (_) { return 4; });
addCalculator(SchemaUtils_1.FieldType.Double, function (_) { return 8; });
addCalculator(SchemaUtils_1.FieldType.ByteArray, function (value) { return value ? value.byteLength : 0; });
addCalculator(SchemaUtils_1.FieldType.Date, function (_) { return 8; });
addCalculator(SchemaUtils_1.FieldType.Text, function (text) {
    return calculateStringUtf8Length(text);
});
exports.Validators = {};
function addValidator(type, validator) {
    exports.Validators[type] = validator;
}
function isNumber(obj) {
    return typeof obj === 'number' && !isNaN(obj);
}
addValidator(SchemaUtils_1.FieldType.Byte, function (value) {
    return !((!Number.isSafeInteger(value)) || value < BYTE_MIN_VALUE || value > BYTE_MAX_VALUE);
});
addValidator(SchemaUtils_1.FieldType.Short, function (value) {
    return !((!Number.isSafeInteger(value)) || value < SHORT_MIN_VALUE || value > SHORT_MAX_VALUE);
});
addValidator(SchemaUtils_1.FieldType.Integer, function (value) {
    return !((!Number.isInteger(value)) || value < INTEGER_MIN_VALUE || value > INTEGER_MAX_VALUE);
});
addValidator(SchemaUtils_1.FieldType.Long, function (value) {
    // Long remaining
    return true;
});
addValidator(SchemaUtils_1.FieldType.Float, function (value) {
    if (!isNumber(value)) {
        return false;
    }
    return !(isNaN(value) || value < FLOAT_MIN_VALUE || value > FLOAT_MAX_VALUE);
});
addValidator(SchemaUtils_1.FieldType.Double, function (value) {
    return !(typeof value !== 'number' || !Number.isFinite(value));
});
addValidator(SchemaUtils_1.FieldType.String, function (value) {
    return typeof value === 'string' && value.length <= MAX_STRING_LENGTH;
});
addValidator(SchemaUtils_1.FieldType.Text, function (value) {
    return typeof value === 'string';
});
addValidator(SchemaUtils_1.FieldType.Boolean, function (value) {
    return typeof value === 'boolean';
});
addValidator(SchemaUtils_1.FieldType.ByteArray, function (value) {
    return value instanceof Uint8Array;
});
addValidator(SchemaUtils_1.FieldType.Date, function (value) {
    return value instanceof Date && !isNaN(value.getTime());
});
/**
 * Processing data verification
 *
 * @author fangpeiwen [fangpeiwen@huawei.com]
 * @since 2020-11-30
 */
var DataModelHelper = /** @class */ (function () {
    function DataModelHelper() {
    }
    DataModelHelper.calculateObject = function (object) {
        var e_1, _a;
        if (!Utils_1.Utils.isNotNullObject(object)) {
            throw new TypeError('object is not valid data model object!');
        }
        var fieldTypeMap = object.getFieldTypeMap();
        var properties = Object.getOwnPropertyNames(object);
        var memorySize = 0;
        try {
            for (var properties_1 = __values(properties), properties_1_1 = properties_1.next(); !properties_1_1.done; properties_1_1 = properties_1.next()) {
                var key = properties_1_1.value;
                var fieldType = SchemaUtils_1.FieldType.getFieldType(fieldTypeMap.get(key));
                var size = exports.Calculators[fieldType] ? exports.Calculators[fieldType](object[key])
                    : ('' + object[key]).length;
                memorySize += size;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (properties_1_1 && !properties_1_1.done && (_a = properties_1.return)) _a.call(properties_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return memorySize + EMPTY_OBJECT_MEMORY_USAGE;
    };
    DataModelHelper.validate = function (object) {
        var e_2, _a;
        if (!Utils_1.Utils.isNotNullObject(object)) {
            throw new TypeError('object is not valid data model object!');
        }
        var properties = Object.getOwnPropertyNames(object);
        var fieldTypeMap = object.getFieldTypeMap();
        if (Utils_1.Utils.isNullOrUndefined(properties) || Utils_1.Utils.isNullOrUndefined(fieldTypeMap)) {
            return false;
        }
        try {
            for (var properties_2 = __values(properties), properties_2_1 = properties_2.next(); !properties_2_1.done; properties_2_1 = properties_2.next()) {
                var name = properties_2_1.value;
                var typeStr = fieldTypeMap.get(name);
                if (!typeStr) {
                    return false;
                }
                var value = object[name];
                if (value === undefined || value === null) {
                    continue;
                }
                if (!exports.Validators[SchemaUtils_1.FieldType.getFieldType(typeStr)](value)) {
                    return false;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (properties_2_1 && !properties_2_1.done && (_a = properties_2.return)) _a.call(properties_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return true;
    };
    DataModelHelper.validateFieldValue = function (fieldType, value) {
        return exports.Validators[SchemaUtils_1.FieldType.getFieldType(fieldType)](value);
    };
    return DataModelHelper;
}());
exports.DataModelHelper = DataModelHelper;
